Fixed AudioManager, classics Sounds and SongNameThread

fabricecaruso opened the pull request:
There were 3 potential crash sources :

Rewritted some parts of AudioManager to avoid init of SDL_Mixer ( with Mix_OpenAudio ) only when the music starts and close it when music stops, and make better cleanings. Init of SDL_Mixer is in Audiomanagers's init, so we can use SDL_Mixer for classic sounds too.

Rewritted Sound class to use SDL_Mixer instead of using SDL_Audio that was disabled.
I don't know if it could lead to **crashes** on linux, but it is still possible as the required audio engine is not running.

Deleted SongNameThread, I moved it to Window::update ( in the main UI loop ) because :

- When you launch a game, the thread is still running, the AudioManager has been deinit and set to null. The thread reinits the AudioManager after 3 seconds. When you come back from the game, Audiomanager is Inited again.
- Accessing songName variable in a thread and also in the UI thread make strings written at the same time and currupts adresses in memory - this can lead to a **crash** with access violation.
- Window::update is never called when a game is launched because system() is waiting for the running process to end.

I also removed this test to move the test in Sound.cpp
if (sInstance == nullptr) // !Settings::getInstance()->getBool("EnableSounds"))
If EnableSounds is false, AudioManager is always null. Talking to Audiomanager to start music makes an **INSTANT CRASH** ( null pointer exception )

Fixed sometimes indentation at 2 spaces with tabs, which is not the original convention in ES ( makes a lot of lines dirty )

tcamargo commented on es-app/CMakeLists.txt:
```diff
@@ -93,8 +92,7 @@ set(ES_SOURCES
     ${CMAKE_CURRENT_SOURCE_DIR}/src/FileFilterIndex.cpp
     ${CMAKE_CURRENT_SOURCE_DIR}/src/SystemScreenSaver.cpp
     ${CMAKE_CURRENT_SOURCE_DIR}/src/CollectionSystemManager.cpp
-    ${CMAKE_CURRENT_SOURCE_DIR}/src/NetworkThread.cpp # batocera
```
Why do you removed NetworkThread?

fabricecaruso replied:
No it's still there

tcamargo replied:
Look at the diff. You removed it from CMakeLists.

fabricecaruso replied:
NO !!! It's still there !!!!!!!!!!! Don't know why git consider it a change.

tcamargo replied:
So, I can't merge you change. It would remove the line..

tcamargo commented on es-core/src/Window.cpp:
```diff
@@ -207,12 +208,21 @@ void Window::input(InputConfig* config, Input input)
-        // batocera
-        if(!mMessages.empty()){
+	// batocera        
+	if (SystemConf::getInstance()->get("audio.display_titles") == "1")
```
What about using mMessages?

fabricecaruso replied:
Still there

tcamargo replied:
Sorry?

fabricecaruso replied:
Still there, just moved after testing songname

tcamargo replied:
yes... why dont use use it? fill mMessages? so, you just need to fill it with the songname and no changes are necessary here.

fabricecaruso replied:
I didn't change previous lines of code found in SongNameThread, just moved them there

tcamargo replied:
And I think it is not the right place. We should have used mMessages, if possible.

tcamargo commented on es-core/src/Sound.h:
```diff
@@ -2,7 +2,7 @@
 #define ES_CORE_SOUND_H
 
-#include "SDL_audio.h"
+#include "SDL_mixer.h"
```
Too many changes in standard ES core code... Why?

fabricecaruso replied:
Sound is a rewrite because It uses the SDL_Audio which is useless ( and don't work anymore ) -> I rewrited it to use SDL_Mixer instead as Music is using it too

tcamargo commented:
@fabricecaruso Could you please detail a bit what are you correcting? I put some comments directly in the code.

fabricecaruso commented:
> @fabricecaruso Could you please detail a bit what are you correcting? I put some comments directly in the code.

Rewritted some parts of AudioManager to avoid ini of SDL_Mixer ( with Mix_OpenAudio ) only when the music starts and close it when it stops, and make better cleanings.
Rewritted Sound class to use SDL_Mixer instead of using SDL_Audio that was disabled.

Deleted SongNameThread, I moved to processing in Window::update ( in the main UI loop ) because :
- When you launch a game, the thread is still running, the AudioManager has been deinit and set to null. The thread reinits the AudioManager after 3 seconds. When you come back from the game, Audiomanager is Inited again.
- Accessing songName variable in a thread and also in the UI thread make strings written at the same time and currupts adresses in memory - this can lead to an access violation.
- Window::update is never called when a game is launched because system() is waiting for the running process to end.

nadenislamarre commented:
i merge.
i build a version with it.
if it causes issues, we will revert.
